src/node-capnp/capnp.cc

```diff
@@ -2521,6 +2521,173 @@ void throw_(const v8::FunctionCallbackInfo<v8::Value>& args) {
+
+  switch (elementSize) {
+    case capnp::ElementSize::VOID:
+    case capnp::ElementSize::BIT:
+    case capnp::ElementSize::BYTE:
+    case capnp::ElementSize::TWO_BYTES:
+    case capnp::ElementSize::FOUR_BYTES:
+    case capnp::ElementSize::EIGHT_BYTES:
+      return query == candidate;
+
+    case capnp::ElementSize::POINTER: {
+      auto queryPtrs = query.as<capnp::List<capnp::AnyPointer>>();
+      auto candidatePtrs = candidate.as<capnp::List<capnp::AnyPointer>>();
+
+      if (queryPtrs.size() != candidatePtrs.size()) return false;
+      for (auto i: kj::indices(queryPtrs)) {
+        if (!matchPowerboxQuery(queryPtrs[i], candidatePtrs[i])) return false;
+      }
+      return true;
+    }
+
+    case capnp::ElementSize::INLINE_COMPOSITE: {
+      // Every element in the query must be matched by at least one element in the candidate.
+      //
+      // TODO(perf): O(m*n), can we do better?
+
+      auto candidateStructs = candidate.as<capnp::List<capnp::AnyStruct>>();
+      for (auto queryElement: query.as<capnp::List<capnp::AnyStruct>>()) {
+        bool matched = false;
+        for (auto candidateElement: candidateStructs) {
+          if (matchPowerboxQuery(queryElement, candidateElement)) {
+            matched = true;
+            break;
+          }
+        }
+        if (!matched) return false;
+      }
+      return true;
+    }
+  }
+}
+
+bool matchPowerboxQuery(capnp::AnyPointer::Reader query, capnp::AnyPointer::Reader candidate) {
+  auto queryType = query.getPointerType();
+  auto candidateType = candidate.getPointerType();
+
+  if (queryType != candidateType) {
+    // Different types -> no match, unless one is null, which is a wildcard.
+    return queryType == capnp::PointerType::NULL_ ||
+           candidateType == capnp::PointerType::NULL_;
+  }
+
+  switch (queryType) {
+    case capnp::PointerType::NULL_:
+      return true;
+    case capnp::PointerType::STRUCT:
+      return matchPowerboxQuery(query.getAs<capnp::AnyStruct>(),
+                                candidate.getAs<capnp::AnyStruct>());
+    case capnp::PointerType::LIST:
+      return matchPowerboxQuery(query.getAs<capnp::AnyList>(),
+                                candidate.getAs<capnp::AnyList>());
+    case capnp::PointerType::CAPABILITY:
+      // TODO(someday): Support matching capabilities?
+      return false;
+  }
+}
+
+void matchPowerboxQuery(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  // matchPowerboxQuery(queryBuffer, candidateBuffer) -> bool
+  //
+  // Decodes queryBuffer and candidateBuffer as capnp messages and returns true if the query
+  // "matches" the candidate according to Sandstorm's powerbox query tag matching algorithm. See
+  // PowerboxDescriptor::Tag::value in sandstorm/powerbox.capnp for full details of the algorithm.
+  // The idea here is that the two parameters are AnyPointer values from Tag::value, which in
+  // node-capnp are normally represented as byte arrays.
```

Comment:
Looks like the only place the matching algorithm is *not* symmetric is in its handling of struct lists. Is that correct?